<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lieven Clement" />


<title>4. Sparse Singular Value Decomposition</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
#rmd-source-code {
  display: none;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">HDA2020</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="intro.html">1. Introduction</a>
    </li>
    <li>
      <a href="svd.html">2. Singular Value Decomposition</a>
    </li>
    <li>
      <a href="prediction.html">3. Prediction with High Dimensional Predictors</a>
    </li>
    <li>
      <a href="sparseSvd.html">4. Sparse Singular Value Decomposition</a>
    </li>
    <li>
      <a href="lda.html">5. Linear Discriminant Analysis</a>
    </li>
    <li>
      <a href="hclust.html">6. Introduction to Hierarchical Clustering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Practicals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Lab1-Intro-SVD.html">Lab 1</a>
    </li>
    <li>
      <a href="Lab2-PCA.html">Lab 2</a>
    </li>
    <li>
      <a href="Lab3-Penalized-Regression.html">Lab 3</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/statOmics/HDA2020">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://statomics.github.io/">statOmics</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">4. Sparse Singular Value Decomposition</h1>
<h4 class="author">Lieven Clement</h4>
<h4 class="date">statOmics, Ghent University (<a href="https://statomics.github.io" class="uri">https://statomics.github.io</a>)</h4>

</div>


<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>In high dimensional data the PCs may have unclear interpretation because they depend on all <span class="math inline">\(p\)</span> original variables.</p>
<div id="example-toxicogenomics-in-early-drug-development" class="section level2">
<h2><span class="header-section-number">1.1</span> Example: Toxicogenomics in early drug development</h2>
<div id="background" class="section level3">
<h3><span class="header-section-number">1.1.1</span> Background</h3>
<ul>
<li><p>Effect of compound on gene expression.</p></li>
<li><p>Insight in action and toxicity of drug in early phase</p></li>
<li><p>Determine activity with bio-assay: e.g. binding afinity of compound to cellwall recepter (target, IC50).</p></li>
<li><p>Early phase: 20 to 50 compounds</p></li>
<li><p>Based on in vitro results one aims to get insight in how to build better compound (higher on-target activity less toxicity.</p></li>
<li><p>Small variantions in moelculair structure lead to variations in BA and gene expression.</p></li>
<li><p>Aim: Build model to predict bio-activity based on gene expression in liver cell line.</p></li>
</ul>
</div>
<div id="data" class="section level3">
<h3><span class="header-section-number">1.1.2</span> Data</h3>
<ul>
<li><p>30 chemical compounds have been screened for toxicity</p></li>
<li><p>Bioassay data on toxicity screening</p></li>
<li><p>Gene expressions in a liver cell line are profiled for each compound (4000 genes)</p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>toxData &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;https://raw.githubusercontent.com/statOmics/HDA2020/data/toxDataCentered.csv&quot;</span>)</span></code></pre></div>
<pre><code>## 
## [36m──[39m [1m[1mColumn specification[1m[22m [36m────────────────────────────────────────────────────────[39m
## cols(
##   .default = col_double()
## )
## [36mℹ[39m Use [30m[47m[30m[47m`spec()`[47m[30m[49m[39m for the full column specifications.</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>svdX &lt;-<span class="st"> </span><span class="kw">svd</span>(toxData[,<span class="op">-</span><span class="dv">1</span>])</span></code></pre></div>
<p>Data is already centered:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>toxData <span class="op">%&gt;%</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="st">  </span>colMeans <span class="op">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="st">  </span>range</span></code></pre></div>
<pre><code>## [1] -1.434038e-17  2.567391e-17</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a> toxData <span class="op">%&gt;%</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="st">  </span>names <span class="op">%&gt;%</span></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="st">  </span>head</span></code></pre></div>
<pre><code>## [1] &quot;BA&quot; &quot;X1&quot; &quot;X2&quot; &quot;X3&quot; &quot;X4&quot; &quot;X5&quot;</code></pre>
<ul>
<li>First column contains data on Bioassay.</li>
<li>The higher the score on Bioassay the more toxic the compound</li>
<li>Other columns contain data on gene expression X1, … , X4000</li>
</ul>
</div>
<div id="data-exploration" class="section level3">
<h3><span class="header-section-number">1.1.3</span> Data exploration</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>toxData <span class="op">%&gt;%</span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span><span class="st">&quot;&quot;</span>,<span class="dt">y=</span>BA)) <span class="op">+</span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="st">  </span><span class="kw">geom_boxplot</span>(<span class="dt">outlier.shape=</span><span class="ot">NA</span>) <span class="op">+</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">position=</span><span class="st">&quot;jitter&quot;</span>)</span></code></pre></div>
<p><img src="sparseSvd_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>svdX &lt;-<span class="st"> </span>toxData[,<span class="op">-</span><span class="dv">1</span>] <span class="op">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="st">  </span>svd</span>
<span id="cb9-3"><a href="#cb9-3"></a></span>
<span id="cb9-4"><a href="#cb9-4"></a>k &lt;-<span class="st"> </span><span class="dv">2</span></span>
<span id="cb9-5"><a href="#cb9-5"></a>Vk &lt;-<span class="st"> </span>svdX<span class="op">$</span>v[,<span class="dv">1</span><span class="op">:</span>k]</span>
<span id="cb9-6"><a href="#cb9-6"></a>Uk &lt;-<span class="st"> </span>svdX<span class="op">$</span>u[,<span class="dv">1</span><span class="op">:</span>k]</span>
<span id="cb9-7"><a href="#cb9-7"></a>Dk &lt;-<span class="st"> </span><span class="kw">diag</span>(svdX<span class="op">$</span>d[<span class="dv">1</span><span class="op">:</span>k])</span>
<span id="cb9-8"><a href="#cb9-8"></a>Zk &lt;-<span class="st"> </span>Uk<span class="op">%*%</span>Dk</span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="kw">colnames</span>(Zk) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;Z&quot;</span>,<span class="dv">1</span><span class="op">:</span>k)</span>
<span id="cb9-10"><a href="#cb9-10"></a><span class="kw">colnames</span>(Vk) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;V&quot;</span>,<span class="dv">1</span><span class="op">:</span>k)</span>
<span id="cb9-11"><a href="#cb9-11"></a></span>
<span id="cb9-12"><a href="#cb9-12"></a>pca &lt;-<span class="st"> </span>Zk <span class="op">%&gt;%</span></span>
<span id="cb9-13"><a href="#cb9-13"></a><span class="st">  </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb9-14"><a href="#cb9-14"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">BA =</span> toxData <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(BA)) <span class="op">%&gt;%</span></span>
<span id="cb9-15"><a href="#cb9-15"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span> Z1, <span class="dt">y =</span> Z2, <span class="dt">color =</span> BA)) <span class="op">+</span></span>
<span id="cb9-16"><a href="#cb9-16"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb9-17"><a href="#cb9-17"></a><span class="st">  </span><span class="kw">scale_colour_gradient2</span>(<span class="dt">low =</span> <span class="st">&quot;blue&quot;</span>,<span class="dt">mid=</span><span class="st">&quot;white&quot;</span>,<span class="dt">high=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span></span>
<span id="cb9-18"><a href="#cb9-18"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">pch =</span> <span class="dv">21</span>, <span class="dt">color =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb9-19"><a href="#cb9-19"></a>pca</span></code></pre></div>
<p><img src="sparseSvd_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<ul>
<li>Scores on the first two principal components (or MDS plot). - Each point corresponds to a compound.</li>
<li>Color code refers to the toxicity score (higher score more toxic).</li>
<li>Clear separation between compounds according to toxicity.</li>
</ul>
<hr />
<ul>
<li>Next logic step in a PCA is to interpret the principal components.</li>
<li>We thus have to assess the loadings.</li>
<li>We can add a vector for each gene to get a biplot, but this would require plotting 4000 vectors, which would render the plot unreadable.</li>
</ul>
<p>Alternative graph to look at the many loadings of the first two PCs.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw">grid.arrange</span>(</span>
<span id="cb10-2"><a href="#cb10-2"></a>  Vk <span class="op">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="st">    </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">geneID =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(Vk)) <span class="op">%&gt;%</span></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> geneID, <span class="dt">y =</span> V1)) <span class="op">+</span></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">pch=</span><span class="dv">21</span>) <span class="op">+</span></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">2</span>)<span class="op">*</span><span class="kw">sd</span>(Vk[,<span class="dv">1</span>]), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>) ,</span>
<span id="cb10-8"><a href="#cb10-8"></a>  Vk <span class="op">%&gt;%</span></span>
<span id="cb10-9"><a href="#cb10-9"></a><span class="st">    </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">geneID =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(Vk)) <span class="op">%&gt;%</span></span>
<span id="cb10-11"><a href="#cb10-11"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> geneID, <span class="dt">y =</span> V2)) <span class="op">+</span></span>
<span id="cb10-12"><a href="#cb10-12"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">pch=</span><span class="dv">21</span>) <span class="op">+</span></span>
<span id="cb10-13"><a href="#cb10-13"></a><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">2</span>)<span class="op">*</span><span class="kw">sd</span>(Vk[,<span class="dv">2</span>]), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>),</span>
<span id="cb10-14"><a href="#cb10-14"></a>  <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="sparseSvd_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<ul>
<li><p>It is almost impossible to interpret the PCs because there are 4000 genes contributing to each PC.</p></li>
<li><p>In an attempt to find the most important genes (in the sense that they drive the interpretation of the PCs), the plots show horizontal reference lines: the average of the loadings, and the average ± twice the standard deviation of the loadings. In between the lines we expects about 95% of the loadings (if they were normally distributed).</p></li>
<li><p>The points outside the band come from the genes that have rather large loadings (in absolute value) and hence are important for the interpretation of the PCs.</p></li>
<li><p>Note, that particularly for the first PC, only a few genes show a markedly large loadings that are negative. This means that an upregulation of these genes will lead to low scores on PC1.</p></li>
<li><p>These genes will very likely play an important role in the toxicity mechanism.</p></li>
<li><p>Indeed, low scores on PC1 are in the direction of more toxicity.</p></li>
</ul>
<hr />
</div>
</div>
<div id="sparse-matrix-decomposition" class="section level2">
<h2><span class="header-section-number">1.2</span> Sparse matrix decomposition</h2>
<p>Basic idea of sparse matrix decomposition of an <span class="math inline">\(n \times p\)</span> matrix <span class="math inline">\(\mathbf{X}\)</span>}</p>
<p>Find a rank <span class="math inline">\(k\)</span> approximation of <span class="math inline">\(\mathbf{X}\)</span> of the form <span class="math display">\[
\mathbf{X}_k = \sum_{j=1}^k \delta_j \mathbf{u}_j\mathbf{v}_j^t
\]</span> such that many of the elements in the <span class="math inline">\(\mathbf{v}_j\)</span> and or <span class="math inline">\(\mathbf{u}_j\)</span> are exactly zero.</p>
<p>We focus on the <strong>Penalised Matrix Decomposition</strong> (PMD) method of Witten <em>et al.</em> (2009, Biostatistics, 10, 3, 515-534).</p>
<p>When many of the loadings in <span class="math inline">\(\mathbf{v}_j\)</span> are zero, the interpretation of the PC only depends on the features that correspond to the remaining non-zero loadings.</p>
<hr />
</div>
</div>
<div id="penalised-matrix-decomposition" class="section level1">
<h1><span class="header-section-number">2</span> Penalised Matrix Decomposition</h1>
<div id="rank-1-approximation" class="section level2">
<h2><span class="header-section-number">2.1</span> Rank 1 approximation</h2>
<p>The rank-1 PMD approximation of the <span class="math inline">\(n \times p\)</span> matrix <span class="math inline">\(\mathbf{X}\)</span> is of the form <span class="math display">\[
\mathbf{X}_1 = \delta_1\mathbf{u}_1\mathbf{v}_1^t,
\]</span></p>
<p>and it is the solution to the following minimisation problem: <span class="math display">\[
\min_{\delta,u,v} \Vert \mathbf{X} - \delta\mathbf{u}\mathbf{v}^t\Vert_F^2 \text{ subject to }
\delta \geq 0, \Vert \mathbf{u}\Vert_2^2=\Vert \mathbf{v}\Vert_2^2=1, \Vert\mathbf{u}\Vert_1\leq c_1, \Vert \mathbf{v}\Vert_1 \leq c_2
\]</span></p>
<p>for some user-defined constants <span class="math inline">\(c_1,c_2&gt;0\)</span>.</p>
<p>The constraints <span class="math inline">\(\Vert\mathbf{u}\Vert_1\leq c_1\)</span> and <span class="math inline">\(\Vert \mathbf{v}\Vert_1 \leq c_2\)</span> are known as <span class="math inline">\(L_1\)</span> or lasso constraints.</p>
<p>Note, <span class="math display">\[
\Vert\mathbf{u}\Vert_1\ = \sum_{i=1}^n \vert u_i \vert \text{ and }\Vert\mathbf{v}\Vert_1\ = \sum_{i=1}^p \vert v_i\vert.
\]</span></p>
<p>If <span class="math inline">\(c_1=+\infty\)</span>, then the constraint on <span class="math inline">\(\mathbf{u}\)</span> is removed (similar for <span class="math inline">\(c_2\)</span>).</p>
<hr />
<p>Note that <span class="math display">\[
  \min_{\delta,u,v} \Vert \mathbf{X} - \delta\mathbf{u}\mathbf{v}^t\Vert_F^2 = \min_{A:\text{rank}(A)=1}  \Vert \mathbf{X} - A\Vert_F^2
\]</span> with <span class="math inline">\(A=X_1\)</span> (truncated SVD of <span class="math inline">\(\mathbf{X}\)</span>, truncated after 1 term). However, the PMD requires a constrained minimisation. The unconstrained solution arises with <span class="math inline">\(c_1=c_2=+\infty\)</span>.</p>
<p>Note that <span class="math inline">\(\Vert \mathbf{u}\Vert_2^2=\Vert \mathbf{v}\Vert_2^2=1\)</span> may also be written as <span class="math inline">\(\mathbf{u}^t\mathbf{u}=\mathbf{v}^t\mathbf{v}=1\)</span> (i.e. the vectors are normalised, just like for the singular vectors of the SVD).</p>
<ul>
<li><p>We also applied the <span class="math inline">\(L_1\)</span> constraint to the least-squares estimator of the regression parameters in a linear regression model.</p></li>
<li><p>Effect was that many of the parameter estimates are set exactly to zero.</p></li>
<li><p>Here, the same effect will take place. Many of the elements in <span class="math inline">\(\mathbf{u}\)</span> and <span class="math inline">\(\mathbf{v}\)</span> will be set exactly to zero.</p></li>
<li><p>In particular, the smaller <span class="math inline">\(c_2\)</span>, the more elements of <span class="math inline">\(\mathbf{v}\)</span> will be set to zero (shrinkage).</p></li>
<li><p>Similarly, the smaller <span class="math inline">\(c_1\)</span>, the more elements of <span class="math inline">\(\mathbf{u}\)</span> will be set to zero.</p></li>
<li><p>The choice of <span class="math inline">\(c_1\)</span> and <span class="math inline">\(c_2\)</span> depends on the objective of the data analysis and will be discussed later.</p></li>
</ul>
<p>The solution is known as a rank-1 PMD approximation of <span class="math inline">\(\mathbf{X}\)</span> because <span class="math inline">\(\delta_1\mathbf{u}_1\mathbf{v}_1^t\)</span> is a matrix of rank 1.</p>
<hr />
</div>
<div id="rank-2-approximation" class="section level2">
<h2><span class="header-section-number">2.2</span> Rank-2 approximation</h2>
<p>The rank-2 PMD approximation is found as follows:</p>
<ol style="list-style-type: decimal">
<li><p>find the rank-1 PMD as before: <span class="math inline">\(\mathbf{X}_1=\delta_1\mathbf{u}_1\mathbf{v}_1^t\)</span>;</p></li>
<li><p>compute <span class="math display">\[
\tilde{\mathbf{X}}_2 = \mathbf{X}-\mathbf{X}_1,
\]</span></p></li>
</ol>
<p>and compute the rank-1 PMD of <span class="math inline">\(\tilde{\mathbf{X}}_2\)</span>, which gives <span class="math inline">\(\delta_2, \mathbf{u}_2, \mathbf{v}_2\)</span>;</p>
<ol start="3" style="list-style-type: decimal">
<li>the rank-2 PMD of <span class="math inline">\(\mathbf{X}\)</span> is then given by <span class="math display">\[
\mathbf{X}_2 = \delta_1\mathbf{u}_1\mathbf{v}_1^t+\delta_2\mathbf{u}_2\mathbf{v}_2^t.
\]</span></li>
</ol>
<p>Note that the <span class="math inline">\(\mathbf{u}\)</span>’s and <span class="math inline">\(\mathbf{v}\)</span>’s are not necessarily orthogonal (only if <span class="math inline">\(c_1, c_2=+\infty\)</span>).</p>
<hr />
</div>
<div id="rank-k-approximation" class="section level2">
<h2><span class="header-section-number">2.3</span> Rank-k approximation</h2>
<p>The rank-<span class="math inline">\(k\)</span> PMD approximation is found as follows:</p>
<ol style="list-style-type: decimal">
<li><p>find the rank-1 PMD as before: <span class="math inline">\(\tilde{\mathbf{X}}_1=\delta_1\mathbf{u}_1\mathbf{v}_1^t\)</span>;</p></li>
<li><p>set <span class="math inline">\(j=1\)</span>;</p></li>
<li><p>compute <span class="math display">\[
\tilde{\mathbf{X}}_{j+1} = \mathbf{X}-\tilde{\mathbf{X}}_1-\cdots - \tilde{\mathbf{X}}_j,
\]</span> and compute the rank-1 PMD of <span class="math inline">\(\tilde{\mathbf{X}}_{j+1}\)</span>, which gives <span class="math inline">\(\delta_{j+1}, \mathbf{u}_{j+1}, \mathbf{v}_{j+1}\)</span>;</p></li>
<li><p>if <span class="math inline">\(j&lt;k\)</span>, increase <span class="math inline">\(j\)</span> with one and go back to step 3; otherwise go to step 5;</p></li>
<li><p>the rank-<span class="math inline">\(k\)</span> PMD of <span class="math inline">\(\mathbf{X}\)</span> is then given by <span class="math display">\[
  \mathbf{X}_k = \sum_{j=1}^k \delta_j\mathbf{u}_j\mathbf{v}_j^t.
\]</span></p></li>
</ol>
<p>Without the <span class="math inline">\(L_1\)</span> constraints, this algorithm gives the ordinary SVD.</p>
<p>As before the <span class="math inline">\(\delta_j\)</span>s, <span class="math inline">\(\mathbf{u}_j\)</span>s and <span class="math inline">\(\mathbf{v}_j\)</span>s are stacked into matrices <span class="math inline">\(\mathbf{D}_k\)</span>, <span class="math inline">\(\mathbf{U}_k\)</span> and <span class="math inline">\(\mathbf{V}_k\)</span>: <span class="math inline">\(\mathbf{X}_k = \mathbf{U}_k\mathbf{D}_k\mathbf{V}_k^t\)</span>.</p>
<p>Note that the <span class="math inline">\(\mathbf{u}\)</span> vectors are generally not orthogonal (i.e. <span class="math inline">\(\mathbf{u}_i^t\mathbf{u}_j\neq 0\)</span> for <span class="math inline">\(i\neq j\)</span>) (the same holds for the <span class="math inline">\(\mathbf{v}\)</span> vectors).</p>
<hr />
</div>
<div id="sparse-pca-via-pmd" class="section level2">
<h2><span class="header-section-number">2.4</span> Sparse PCA via PMD</h2>
<p>Just like the SVD can be used to find the PCA solution, the PMD can be used to find a sparse PCA solution (SPC).</p>
<p>For a PCA, it is only important to have sparse loadings (in the <span class="math inline">\(\mathbf{v}\)</span>s); the scores (involving the <span class="math inline">\(\mathbf{u}\)</span>s) need not be sparse.</p>
<p>The rank-1 solutions thus satisfy <span class="math display">\[
  \min_{\delta,u,v} \Vert \mathbf{X} - \delta\mathbf{u}\mathbf{v}^t\Vert_F^2 \text{ subject to }
  \delta \geq 0, \Vert \mathbf{u}\Vert_2^2=\Vert \mathbf{v}\Vert_2^2=1, \vert \mathbf{v}\vert \leq c
\]</span> (i.e. no sparsity constraint on <span class="math inline">\(\mathbf{u}\)</span>).</p>
<p>Hence the PCA scores <span class="math display">\[
 \mathbf{Z}_k = \mathbf{U}_k\mathbf{D}_k
\]</span> relate to <span class="math inline">\(\mathbf{X}_k\)</span> through the sparse loadings in <span class="math inline">\(\mathbf{V}_k\)</span>.</p>
<hr />
<p>It can be shown that the SPC solution (rank-1) also maximises <span class="math display">\[
  \mathbf{v}^t\mathbf{X}^t\mathbf{X}\mathbf{v} \text{ subject to }
  \Vert \mathbf{u}\Vert_2^2=\Vert \mathbf{v}\Vert_2^2=1, \Vert \mathbf{v}\Vert_1 \leq c
\]</span> (i.e. it maximises the variance of the PCs, subject to sparsity constraint on loadings).</p>
<p>From these two slides we may conclude that we can use the PMD as the basis of a sparse PCA, and that the interpretation is as before in terms of maximising variance. Only now we hope that only a few features will give a non-zero loading in the <span class="math inline">\(\mathbf{v}\)</span>-vectors. Biplots may again be constructed (with fewer vectors to be plotted).</p>
<hr />
</div>
</div>
<div id="toxicogenomics-analysis-with-a-sparse-pca" class="section level1">
<h1><span class="header-section-number">3</span> Toxicogenomics analysis with a sparse PCA</h1>
<div id="spca" class="section level2">
<h2><span class="header-section-number">3.1</span> SPCA</h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">library</span>(PMA)</span>
<span id="cb11-2"><a href="#cb11-2"></a></span>
<span id="cb11-3"><a href="#cb11-3"></a>spcX &lt;-<span class="st"> </span><span class="kw">SPC</span>(</span>
<span id="cb11-4"><a href="#cb11-4"></a>  toxData[,<span class="op">-</span><span class="dv">1</span>] <span class="op">%&gt;%</span></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="st">    </span>as.matrix,</span>
<span id="cb11-6"><a href="#cb11-6"></a>  <span class="dt">K=</span><span class="dv">30</span>,</span>
<span id="cb11-7"><a href="#cb11-7"></a>  <span class="dt">sumabsv =</span> <span class="dv">5</span>)</span></code></pre></div>
<pre><code>## 12345678910111213141516171819
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="kw">barplot</span>(spcX<span class="op">$</span>d<span class="op">^</span><span class="dv">2</span>,<span class="dt">ylab=</span><span class="st">&quot;eigenvalue&quot;</span>,<span class="dt">cex.lab=</span><span class="fl">1.5</span>)</span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="kw">barplot</span>(spcX<span class="op">$</span>prop.var.explained,</span>
<span id="cb13-4"><a href="#cb13-4"></a>     <span class="dt">ylab=</span><span class="st">&quot;cummulative variance&quot;</span>,<span class="dt">cex.lab=</span><span class="fl">1.5</span>)</span></code></pre></div>
<p><img src="sparseSvd_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<ul>
<li><p>The plot on the left shows the eigenvalues (squared singular values) from the PMD.</p>
<ul>
<li>This suggests that the first one or two are the most important, but all eigenvalues remain substantially larger than zero.</li>
</ul></li>
<li><p>The figure on the right shows the cumulative percentage of variance retained in the PCs.</p>
<ul>
<li>These percentages are now no longer calculated as before as ratio’s of sums of eigenvalues.</li>
<li>A more complicated calculation is required now, because the PCs are no longer uncorrelated with one another.</li>
<li>Thus part of the information (variance) of one PC is shared with part of the information of another PCs (i.e. non-zero correlation or covariance).</li>
</ul></li>
</ul>
</div>
<div id="loading" class="section level2">
<h2><span class="header-section-number">3.2</span> Loading</h2>
<p>Loadings of the first two PCs, with constraint <span class="math inline">\(\Vert\mathbf{v}\Vert_1 &lt; 5\)</span>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>k &lt;-<span class="st"> </span><span class="dv">2</span></span>
<span id="cb14-2"><a href="#cb14-2"></a>Vk &lt;-<span class="st"> </span>spcX<span class="op">$</span>v[,<span class="dv">1</span><span class="op">:</span>k]</span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="kw">colnames</span>(Vk) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;V&quot;</span>,<span class="dv">1</span><span class="op">:</span>k)</span>
<span id="cb14-4"><a href="#cb14-4"></a></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="kw">grid.arrange</span>(</span>
<span id="cb14-6"><a href="#cb14-6"></a>  Vk <span class="op">%&gt;%</span></span>
<span id="cb14-7"><a href="#cb14-7"></a><span class="st">    </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb14-8"><a href="#cb14-8"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">geneID =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(Vk)) <span class="op">%&gt;%</span></span>
<span id="cb14-9"><a href="#cb14-9"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> geneID, <span class="dt">y =</span> V1)) <span class="op">+</span></span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">pch=</span><span class="dv">21</span>) <span class="op">+</span></span>
<span id="cb14-11"><a href="#cb14-11"></a><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">2</span>)<span class="op">*</span><span class="kw">sd</span>(Vk[,<span class="dv">1</span>]), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>) ,</span>
<span id="cb14-12"><a href="#cb14-12"></a>  Vk <span class="op">%&gt;%</span></span>
<span id="cb14-13"><a href="#cb14-13"></a><span class="st">    </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb14-14"><a href="#cb14-14"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">geneID =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(Vk)) <span class="op">%&gt;%</span></span>
<span id="cb14-15"><a href="#cb14-15"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> geneID, <span class="dt">y =</span> V2)) <span class="op">+</span></span>
<span id="cb14-16"><a href="#cb14-16"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">pch=</span><span class="dv">21</span>) <span class="op">+</span></span>
<span id="cb14-17"><a href="#cb14-17"></a><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">2</span>)<span class="op">*</span><span class="kw">sd</span>(Vk[,<span class="dv">2</span>]), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>),</span>
<span id="cb14-18"><a href="#cb14-18"></a>  <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="sparseSvd_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>These plots show the loadings of the first two sparse PCs. Many of the loadings are now exactly equal to zero. Only few genes show large loadings.</p>
</div>
<div id="change-constraint" class="section level2">
<h2><span class="header-section-number">3.3</span> Change constraint</h2>
<p>We now repeat the analysis, but with the constraint <span class="math inline">\(\Vert \mathbf{v} \Vert_1 &lt; 1\)</span>. This gives a much sparser solution with only two genes with non-zero loadings in each dimension.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>spcX1 &lt;-<span class="st"> </span><span class="kw">SPC</span>(</span>
<span id="cb15-2"><a href="#cb15-2"></a>  toxData[,<span class="op">-</span><span class="dv">1</span>] <span class="op">%&gt;%</span></span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="st">    </span>as.matrix,</span>
<span id="cb15-4"><a href="#cb15-4"></a>  <span class="dt">K=</span><span class="dv">30</span>,</span>
<span id="cb15-5"><a href="#cb15-5"></a>  <span class="dt">sumabsv =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12
## 12</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>k &lt;-<span class="st"> </span><span class="dv">2</span></span>
<span id="cb17-2"><a href="#cb17-2"></a>Vk &lt;-<span class="st"> </span>spcX1<span class="op">$</span>v[,<span class="dv">1</span><span class="op">:</span>k]</span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="kw">colnames</span>(Vk) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;V&quot;</span>, <span class="dv">1</span><span class="op">:</span>k)</span>
<span id="cb17-4"><a href="#cb17-4"></a></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="kw">grid.arrange</span>(</span>
<span id="cb17-6"><a href="#cb17-6"></a>  Vk <span class="op">%&gt;%</span></span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="st">    </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">geneID =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(Vk)) <span class="op">%&gt;%</span></span>
<span id="cb17-9"><a href="#cb17-9"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> geneID, <span class="dt">y =</span> V1)) <span class="op">+</span></span>
<span id="cb17-10"><a href="#cb17-10"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">pch=</span><span class="dv">21</span>) <span class="op">+</span></span>
<span id="cb17-11"><a href="#cb17-11"></a><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">2</span>)<span class="op">*</span><span class="kw">sd</span>(Vk[,<span class="dv">1</span>]), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>) ,</span>
<span id="cb17-12"><a href="#cb17-12"></a>  Vk <span class="op">%&gt;%</span></span>
<span id="cb17-13"><a href="#cb17-13"></a><span class="st">    </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb17-14"><a href="#cb17-14"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">geneID =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(Vk)) <span class="op">%&gt;%</span></span>
<span id="cb17-15"><a href="#cb17-15"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> geneID, <span class="dt">y =</span> V2)) <span class="op">+</span></span>
<span id="cb17-16"><a href="#cb17-16"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">pch=</span><span class="dv">21</span>) <span class="op">+</span></span>
<span id="cb17-17"><a href="#cb17-17"></a><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">2</span>)<span class="op">*</span><span class="kw">sd</span>(Vk[,<span class="dv">2</span>]), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>),</span>
<span id="cb17-18"><a href="#cb17-18"></a>  <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="sparseSvd_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<hr />
<p>We now repeat the analysis, but with the constraint <span class="math inline">\(\Vert \mathbf{v} \Vert_1 &lt; 10\)</span>. This gives a much less sparse solution with several genes with non-zero loadings.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>spcX10 &lt;-<span class="st"> </span><span class="kw">SPC</span>(</span>
<span id="cb18-2"><a href="#cb18-2"></a>  toxData[,<span class="op">-</span><span class="dv">1</span>] <span class="op">%&gt;%</span></span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="st">    </span>as.matrix,</span>
<span id="cb18-4"><a href="#cb18-4"></a>  <span class="dt">K=</span><span class="dv">30</span>,</span>
<span id="cb18-5"><a href="#cb18-5"></a>  <span class="dt">sumabsv =</span> <span class="dv">10</span>)</span></code></pre></div>
<pre><code>## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920
## 1234567891011121314151617181920</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>k &lt;-<span class="st"> </span><span class="dv">2</span></span>
<span id="cb20-2"><a href="#cb20-2"></a>Vk &lt;-<span class="st"> </span>spcX10<span class="op">$</span>v[,<span class="dv">1</span><span class="op">:</span>k]</span>
<span id="cb20-3"><a href="#cb20-3"></a><span class="kw">colnames</span>(Vk) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;V&quot;</span>,<span class="dv">1</span><span class="op">:</span>k)</span>
<span id="cb20-4"><a href="#cb20-4"></a></span>
<span id="cb20-5"><a href="#cb20-5"></a><span class="kw">grid.arrange</span>(</span>
<span id="cb20-6"><a href="#cb20-6"></a>  Vk <span class="op">%&gt;%</span></span>
<span id="cb20-7"><a href="#cb20-7"></a><span class="st">    </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb20-8"><a href="#cb20-8"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">geneID =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(Vk)) <span class="op">%&gt;%</span></span>
<span id="cb20-9"><a href="#cb20-9"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> geneID, <span class="dt">y =</span> V1)) <span class="op">+</span></span>
<span id="cb20-10"><a href="#cb20-10"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">pch=</span><span class="dv">21</span>) <span class="op">+</span></span>
<span id="cb20-11"><a href="#cb20-11"></a><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">2</span>)<span class="op">*</span><span class="kw">sd</span>(Vk[,<span class="dv">1</span>]), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>) ,</span>
<span id="cb20-12"><a href="#cb20-12"></a>  Vk <span class="op">%&gt;%</span></span>
<span id="cb20-13"><a href="#cb20-13"></a><span class="st">    </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb20-14"><a href="#cb20-14"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">geneID =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(Vk)) <span class="op">%&gt;%</span></span>
<span id="cb20-15"><a href="#cb20-15"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> geneID, <span class="dt">y =</span> V2)) <span class="op">+</span></span>
<span id="cb20-16"><a href="#cb20-16"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">pch=</span><span class="dv">21</span>) <span class="op">+</span></span>
<span id="cb20-17"><a href="#cb20-17"></a><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">2</span>)<span class="op">*</span><span class="kw">sd</span>(Vk[,<span class="dv">2</span>]), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>),</span>
<span id="cb20-18"><a href="#cb20-18"></a>  <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="sparseSvd_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<hr />
</div>
<div id="variance-explained" class="section level2">
<h2><span class="header-section-number">3.4</span> Variance explained</h2>
<p>Variance explained for <span class="math inline">\(c=1,5,10\)</span> and <span class="math inline">\(+\infty\)</span>.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="kw">data.frame</span>(</span>
<span id="cb21-2"><a href="#cb21-2"></a>  <span class="dt">PCs =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>,</span>
<span id="cb21-3"><a href="#cb21-3"></a>  <span class="st">&quot;c1&quot;</span> =<span class="st"> </span>spcX1<span class="op">$</span>prop.var.explained,</span>
<span id="cb21-4"><a href="#cb21-4"></a>  <span class="st">&quot;c5&quot;</span> =<span class="st"> </span>spcX<span class="op">$</span>prop.var.explained,</span>
<span id="cb21-5"><a href="#cb21-5"></a>  <span class="st">&quot;c10&quot;</span> =<span class="st"> </span>spcX10<span class="op">$</span>prop.var.explained,</span>
<span id="cb21-6"><a href="#cb21-6"></a>  <span class="st">&quot;conventional&quot;</span> =<span class="st"> </span><span class="kw">cumsum</span>(svdX<span class="op">$</span>d<span class="op">^</span><span class="dv">2</span>)<span class="op">/</span><span class="kw">sum</span>(svdX<span class="op">$</span>d<span class="op">^</span><span class="dv">2</span>)</span>
<span id="cb21-7"><a href="#cb21-7"></a>  ) <span class="op">%&gt;%</span></span>
<span id="cb21-8"><a href="#cb21-8"></a><span class="st">  </span><span class="kw">gather</span>(<span class="st">&quot;method&quot;</span>,<span class="st">&quot;varExpl&quot;</span>,<span class="op">-</span><span class="dv">1</span>) <span class="op">%&gt;%</span></span>
<span id="cb21-9"><a href="#cb21-9"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> PCs, <span class="dt">y =</span> varExpl, <span class="dt">color =</span> method)) <span class="op">+</span></span>
<span id="cb21-10"><a href="#cb21-10"></a><span class="st">  </span><span class="kw">geom_line</span>() <span class="op">+</span></span>
<span id="cb21-11"><a href="#cb21-11"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb21-12"><a href="#cb21-12"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&quot;Variance explained (%)&quot;</span>)</span></code></pre></div>
<p><img src="sparseSvd_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>This graph illustrates that the percentages of variance explained by the PCs are smaller for sparser solutions. Thus in choosing an appropriate penalty parameter <span class="math inline">\(c_2\)</span> one should not only look at the sparseness of the solution, but also at the percentage of variance contained in the first few PCs. If this percentage is very small, then the graphs of the scores (or biplots) are no longer very informative (i.e. they will not tell the full story). Thus a compromise must be looked for.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="kw">library</span>(latex2exp)</span>
<span id="cb22-2"><a href="#cb22-2"></a></span>
<span id="cb22-3"><a href="#cb22-3"></a>Uk5 &lt;-<span class="st"> </span>spcX<span class="op">$</span>u[,<span class="dv">1</span><span class="op">:</span>k]</span>
<span id="cb22-4"><a href="#cb22-4"></a>Dk5 &lt;-<span class="st"> </span><span class="kw">diag</span>(spcX<span class="op">$</span>d[<span class="dv">1</span><span class="op">:</span>k])</span>
<span id="cb22-5"><a href="#cb22-5"></a>Zk5 &lt;-<span class="st"> </span>Uk5<span class="op">%*%</span>Dk5</span>
<span id="cb22-6"><a href="#cb22-6"></a><span class="kw">colnames</span>(Zk5) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;Z&quot;</span>,<span class="dv">1</span><span class="op">:</span>k)</span>
<span id="cb22-7"><a href="#cb22-7"></a></span>
<span id="cb22-8"><a href="#cb22-8"></a>spc5 &lt;-<span class="st"> </span>Zk5 <span class="op">%&gt;%</span></span>
<span id="cb22-9"><a href="#cb22-9"></a><span class="st">  </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb22-10"><a href="#cb22-10"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">BA =</span> toxData <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(BA)) <span class="op">%&gt;%</span></span>
<span id="cb22-11"><a href="#cb22-11"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span> Z1, <span class="dt">y =</span> Z2, <span class="dt">color =</span> BA)) <span class="op">+</span></span>
<span id="cb22-12"><a href="#cb22-12"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb22-13"><a href="#cb22-13"></a><span class="st">  </span><span class="kw">scale_colour_gradient2</span>(<span class="dt">low =</span> <span class="st">&quot;blue&quot;</span>,<span class="dt">mid=</span><span class="st">&quot;white&quot;</span>,<span class="dt">high=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span></span>
<span id="cb22-14"><a href="#cb22-14"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">pch =</span> <span class="dv">21</span>, <span class="dt">color =</span> <span class="st">&quot;black&quot;</span>) <span class="op">+</span></span>
<span id="cb22-15"><a href="#cb22-15"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="kw">TeX</span>(<span class="st">&quot;$c_2 = 5$&quot;</span>))</span>
<span id="cb22-16"><a href="#cb22-16"></a></span>
<span id="cb22-17"><a href="#cb22-17"></a>Uk1 &lt;-<span class="st"> </span>spcX1<span class="op">$</span>u[,<span class="dv">1</span><span class="op">:</span>k]</span>
<span id="cb22-18"><a href="#cb22-18"></a>Dk1 &lt;-<span class="st"> </span><span class="kw">diag</span>(spcX1<span class="op">$</span>d[<span class="dv">1</span><span class="op">:</span>k])</span>
<span id="cb22-19"><a href="#cb22-19"></a>Zk1 &lt;-<span class="st"> </span>Uk1<span class="op">%*%</span>Dk1</span>
<span id="cb22-20"><a href="#cb22-20"></a><span class="kw">colnames</span>(Zk1) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;Z&quot;</span>,<span class="dv">1</span><span class="op">:</span>k)</span>
<span id="cb22-21"><a href="#cb22-21"></a></span>
<span id="cb22-22"><a href="#cb22-22"></a>spc1 &lt;-<span class="st"> </span>Zk1 <span class="op">%&gt;%</span></span>
<span id="cb22-23"><a href="#cb22-23"></a><span class="st">  </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb22-24"><a href="#cb22-24"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">BA =</span> toxData <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(BA)) <span class="op">%&gt;%</span></span>
<span id="cb22-25"><a href="#cb22-25"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span> Z1, <span class="dt">y =</span> Z2, <span class="dt">color =</span> BA)) <span class="op">+</span></span>
<span id="cb22-26"><a href="#cb22-26"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb22-27"><a href="#cb22-27"></a><span class="st">  </span><span class="kw">scale_colour_gradient2</span>(<span class="dt">low =</span> <span class="st">&quot;blue&quot;</span>,<span class="dt">mid=</span><span class="st">&quot;white&quot;</span>,<span class="dt">high=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span></span>
<span id="cb22-28"><a href="#cb22-28"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">pch =</span> <span class="dv">21</span>, <span class="dt">color =</span> <span class="st">&quot;black&quot;</span>) <span class="op">+</span></span>
<span id="cb22-29"><a href="#cb22-29"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="kw">TeX</span>(<span class="st">&quot;$c_2 = 1$&quot;</span>))</span>
<span id="cb22-30"><a href="#cb22-30"></a></span>
<span id="cb22-31"><a href="#cb22-31"></a>Uk10 &lt;-<span class="st"> </span>spcX10<span class="op">$</span>u[,<span class="dv">1</span><span class="op">:</span>k]</span>
<span id="cb22-32"><a href="#cb22-32"></a>Dk10 &lt;-<span class="st"> </span><span class="kw">diag</span>(spcX10<span class="op">$</span>d[<span class="dv">1</span><span class="op">:</span>k])</span>
<span id="cb22-33"><a href="#cb22-33"></a>Zk10 &lt;-<span class="st"> </span>Uk10<span class="op">%*%</span>Dk10</span>
<span id="cb22-34"><a href="#cb22-34"></a><span class="kw">colnames</span>(Zk10) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;Z&quot;</span>,<span class="dv">1</span><span class="op">:</span>k)</span>
<span id="cb22-35"><a href="#cb22-35"></a></span>
<span id="cb22-36"><a href="#cb22-36"></a>spc10 &lt;-<span class="st"> </span>Zk10 <span class="op">%&gt;%</span></span>
<span id="cb22-37"><a href="#cb22-37"></a><span class="st">  </span>as.data.frame <span class="op">%&gt;%</span></span>
<span id="cb22-38"><a href="#cb22-38"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">BA =</span> toxData <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(BA)) <span class="op">%&gt;%</span></span>
<span id="cb22-39"><a href="#cb22-39"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span> Z1, <span class="dt">y =</span> Z2, <span class="dt">color =</span> BA)) <span class="op">+</span></span>
<span id="cb22-40"><a href="#cb22-40"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb22-41"><a href="#cb22-41"></a><span class="st">  </span><span class="kw">scale_colour_gradient2</span>(<span class="dt">low =</span> <span class="st">&quot;blue&quot;</span>,<span class="dt">mid=</span><span class="st">&quot;white&quot;</span>,<span class="dt">high=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span></span>
<span id="cb22-42"><a href="#cb22-42"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">pch =</span> <span class="dv">21</span>, <span class="dt">color =</span> <span class="st">&quot;black&quot;</span>) <span class="op">+</span></span>
<span id="cb22-43"><a href="#cb22-43"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="kw">TeX</span>(<span class="st">&quot;$c_2 = 10$&quot;</span>))</span>
<span id="cb22-44"><a href="#cb22-44"></a></span>
<span id="cb22-45"><a href="#cb22-45"></a><span class="kw">grid.arrange</span>(spc1, spc5, spc10, pca <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="kw">TeX</span>(<span class="st">&quot;$c_2 = </span><span class="ch">\\</span><span class="st">infty&quot;</span>)), <span class="dt">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="sparseSvd_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>Finally we show the plots of the scores on the first two sparse PCs for <span class="math inline">\(c_2=1\)</span> (top-left), <span class="math inline">\(c_2=5\)</span> (top-right) and <span class="math inline">\(c_2=10\)</span> (bottom-left) and <span class="math inline">\(c_2=\infty\)</span> (bottom-right, convention SVD).</p>
<p>Note, that the most sparse solution (top-left) shows a strong positive correlation between the first two PCs (as a consequence of the loss of orthogonality of the singular vectors).</p>
<p>All graphs succeed quite well in separating the toxic from the non-toxic compounds based on their gene expression.</p>
<ul>
<li><p>Hence, we may expect that the two genes identified from the most sparse solution play an important role in the development of toxic effects.</p></li>
<li><p>Note, however, that our data analysis did not provide a scientific proof for this conclusion: (sparse) PCA is only an exploratory or descriptive data analysis method that helps the data-analyst to gain insight into the data.</p></li>
<li><p>The fact that we found a few genes that appear to be associated with toxicity is of course a nice result.</p></li>
<li><p>It is important to understand that the toxicity outcome was not used in the (sparse) PCA. We only used the toxicity to color the points in the graphs. PCA is an unsupervised method.</p></li>
</ul>
</div>
</div>
<div id="acknowledgement" class="section level1 unnumbered">
<h1>Acknowledgement</h1>
<ul>
<li>Olivier Thas for sharing his materials of Analysis of High Dimensional Data 2019-2020, which I used as the starting point for this chapter.</li>
</ul>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1>Session info</h1>
<details>
<p><summary>Session info</summary></p>
<pre><code>## [1] &quot;2020-11-10 00:22:48 UTC&quot;</code></pre>
<pre><code>## ─ Session info ───────────────────────────────────────────────────────────────
##  setting  value                       
##  version  R version 4.0.3 (2020-10-10)
##  os       macOS Catalina 10.15.7      
##  system   x86_64, darwin17.0          
##  ui       X11                         
##  language (EN)                        
##  collate  en_US.UTF-8                 
##  ctype    en_US.UTF-8                 
##  tz       UTC                         
##  date     2020-11-10                  
## 
## ─ Packages ───────────────────────────────────────────────────────────────────
##  package                * version    date       lib
##  AIMS                   * 1.22.0     2020-10-27 [1]
##  amap                     0.8-18     2019-12-12 [1]
##  AnnotationDbi            1.52.0     2020-10-27 [1]
##  AnnotationHub          * 2.22.0     2020-10-27 [1]
##  askpass                  1.1        2019-01-13 [1]
##  assertthat               0.2.1      2019-03-21 [1]
##  backports                1.2.0      2020-11-02 [1]
##  beachmat                 2.6.1      2020-11-06 [1]
##  beeswarm                 0.2.3      2016-04-25 [1]
##  Biobase                * 2.50.0     2020-10-27 [1]
##  BiocFileCache          * 1.14.0     2020-10-27 [1]
##  BiocGenerics           * 0.36.0     2020-10-27 [1]
##  BiocManager              1.30.10    2019-11-16 [1]
##  BiocNeighbors            1.8.0      2020-10-27 [1]
##  BiocParallel             1.24.1     2020-11-06 [1]
##  BiocSingular             1.6.0      2020-10-27 [1]
##  BiocVersion              3.12.0     2020-05-14 [1]
##  biomaRt                * 2.46.0     2020-10-27 [1]
##  bit                      4.0.4      2020-08-04 [1]
##  bit64                    4.0.5      2020-08-30 [1]
##  bitops                   1.0-6      2013-08-17 [1]
##  blob                     1.2.1      2020-01-20 [1]
##  boot                   * 1.3-25     2020-04-26 [1]
##  bootstrap                2019.6     2019-06-17 [1]
##  breastCancerMAINZ      * 1.28.0     2020-10-29 [1]
##  broom                    0.7.2      2020-10-20 [1]
##  callr                    3.5.1      2020-10-13 [1]
##  CCA                    * 1.2        2012-10-29 [1]
##  cellranger               1.1.0      2016-07-27 [1]
##  class                    7.3-17     2020-04-26 [2]
##  cli                      2.1.0      2020-10-12 [1]
##  cluster                * 2.1.0      2019-06-19 [1]
##  codetools                0.2-16     2018-12-24 [2]
##  colorspace               1.4-1      2019-03-18 [1]
##  crayon                   1.3.4      2017-09-16 [1]
##  curl                     4.3        2019-12-02 [1]
##  DAAG                   * 1.24       2020-03-10 [1]
##  DBI                      1.1.0      2019-12-15 [1]
##  dbplyr                 * 2.0.0      2020-11-03 [1]
##  DelayedArray             0.16.0     2020-10-27 [1]
##  DelayedMatrixStats       1.12.0     2020-10-27 [1]
##  desc                     1.2.0      2018-05-01 [1]
##  devtools                 2.3.2      2020-09-18 [1]
##  digest                   0.6.27     2020-10-24 [1]
##  dotCall64              * 1.0-0      2018-07-30 [1]
##  dplyr                  * 1.0.2      2020-08-18 [1]
##  e1071                  * 1.7-4      2020-10-14 [1]
##  elasticnet               1.3        2020-05-15 [1]
##  ellipsis                 0.3.1      2020-05-15 [1]
##  evaluate                 0.14       2019-05-28 [1]
##  ExperimentHub          * 1.16.0     2020-10-27 [1]
##  fansi                    0.4.1      2020-01-08 [1]
##  farver                   2.0.3      2020-01-16 [1]
##  fastmap                  1.0.1      2019-10-08 [1]
##  fda                    * 5.1.5.1    2020-08-02 [1]
##  fields                 * 11.6       2020-10-09 [1]
##  forcats                * 0.5.0      2020-03-01 [1]
##  foreach                  1.5.1      2020-10-15 [1]
##  fs                       1.5.0      2020-07-31 [1]
##  genefu                 * 2.22.0     2020-10-27 [1]
##  generics                 0.1.0      2020-10-31 [1]
##  GenomeInfoDb           * 1.26.0     2020-10-27 [1]
##  GenomeInfoDbData         1.2.4      2020-11-10 [1]
##  GenomicRanges          * 1.42.0     2020-10-27 [1]
##  ggbeeswarm               0.6.0      2017-08-07 [1]
##  ggbiplot               * 0.55       2020-11-10 [1]
##  ggforce                * 0.3.2      2020-06-23 [1]
##  ggplot2                * 3.3.2      2020-06-19 [1]
##  GIGrvg                   0.5        2017-06-10 [1]
##  git2r                    0.27.1     2020-05-03 [1]
##  glmnet                 * 4.0-2      2020-06-16 [1]
##  glue                     1.4.2      2020-08-27 [1]
##  gridExtra              * 2.3        2017-09-09 [1]
##  gtable                   0.3.0      2019-03-25 [1]
##  haven                    2.3.1      2020-06-01 [1]
##  highr                    0.8        2019-03-20 [1]
##  hms                      0.5.3      2020-01-08 [1]
##  htmltools                0.5.0      2020-06-16 [1]
##  httpuv                   1.5.4      2020-06-06 [1]
##  httr                     1.4.2      2020-07-20 [1]
##  HyperbolicDist           0.6-2      2009-09-23 [1]
##  iC10                   * 1.5        2019-02-08 [1]
##  iC10TrainingData       * 1.3.1      2018-08-24 [1]
##  impute                 * 1.64.0     2020-10-27 [1]
##  interactiveDisplayBase   1.28.0     2020-10-27 [1]
##  IRanges                * 2.24.0     2020-10-27 [1]
##  irlba                    2.3.3      2019-02-05 [1]
##  iterators                1.0.13     2020-10-15 [1]
##  jpeg                     0.1-8.1    2019-10-24 [1]
##  jsonlite                 1.7.1      2020-09-07 [1]
##  KernSmooth               2.23-17    2020-04-26 [2]
##  knitr                    1.30       2020-09-22 [1]
##  labeling                 0.4.2      2020-10-20 [1]
##  lars                     1.2        2013-04-24 [1]
##  later                    1.1.0.1    2020-06-05 [1]
##  latex2exp              * 0.4.0      2015-11-30 [1]
##  lattice                * 0.20-41    2020-04-02 [2]
##  latticeExtra             0.6-29     2019-12-19 [1]
##  lava                     1.6.8.1    2020-11-04 [1]
##  lifecycle                0.2.0      2020-03-06 [1]
##  limma                  * 3.46.0     2020-10-27 [1]
##  lubridate                1.7.9      2020-06-08 [1]
##  magrittr                 1.5        2014-11-22 [1]
##  maps                     3.3.0      2018-04-03 [1]
##  MASS                     7.3-53     2020-09-09 [2]
##  Matrix                 * 1.2-18     2019-11-27 [2]
##  MatrixGenerics         * 1.2.0      2020-10-27 [1]
##  matrixStats            * 0.57.0     2020-09-25 [1]
##  mclust                 * 5.4.6      2020-04-11 [1]
##  mda                      0.5-2      2020-06-29 [1]
##  memoise                  1.1.0      2017-04-21 [1]
##  mgcv                   * 1.8-33     2020-08-27 [1]
##  mime                     0.9        2020-02-04 [1]
##  misc3d                   0.9-0      2020-09-06 [1]
##  modelr                   0.1.8      2020-05-19 [1]
##  munsell                  0.5.0      2018-06-12 [1]
##  muscData               * 1.4.0      2020-10-29 [1]
##  nlme                   * 3.1-149    2020-08-23 [2]
##  NormalBetaPrime        * 2.2        2019-01-19 [1]
##  openssl                  1.4.3      2020-09-18 [1]
##  pamr                   * 1.56.1     2019-04-22 [1]
##  pillar                   1.4.6      2020-07-10 [1]
##  pkgbuild                 1.1.0      2020-07-13 [1]
##  pkgconfig                2.0.3      2019-09-22 [1]
##  pkgload                  1.1.0      2020-05-29 [1]
##  plot3D                 * 1.3        2019-12-18 [1]
##  plotROC                * 2.2.1      2018-06-23 [1]
##  pls                    * 2.7-3      2020-08-07 [1]
##  plyr                   * 1.8.6      2020-03-03 [1]
##  PMA                    * 1.2.1      2020-02-03 [1]
##  png                      0.1-7      2013-12-03 [1]
##  polyclip                 1.10-0     2019-03-14 [1]
##  pracma                   2.2.9      2019-12-15 [1]
##  prettyunits              1.1.1      2020-01-24 [1]
##  processx                 3.4.4      2020-09-03 [1]
##  prodlim                * 2019.11.13 2019-11-17 [1]
##  progress                 1.2.2      2019-05-16 [1]
##  promises                 1.1.1      2020-06-09 [1]
##  ps                       1.4.0      2020-10-07 [1]
##  pscl                     1.5.5      2020-03-07 [1]
##  purrr                  * 0.3.4      2020-04-17 [1]
##  R6                       2.5.0      2020-10-28 [1]
##  rappdirs                 0.3.1      2016-03-28 [1]
##  RColorBrewer             1.1-2      2014-12-07 [1]
##  Rcpp                     1.0.5      2020-07-06 [1]
##  RCurl                    1.98-1.2   2020-04-18 [1]
##  readr                  * 1.4.0      2020-10-05 [1]
##  readxl                   1.3.1      2019-03-13 [1]
##  remotes                  2.2.0      2020-07-21 [1]
##  reprex                   0.3.0      2019-05-16 [1]
##  rlang                    0.4.8      2020-10-08 [1]
##  rmarkdown                2.5        2020-10-21 [1]
##  rmeta                    3.0        2018-03-20 [1]
##  rprojroot                1.3-2      2018-01-03 [1]
##  RSQLite                  2.2.1      2020-09-30 [1]
##  rstudioapi               0.11       2020-02-07 [1]
##  rsvd                     1.0.3      2020-02-17 [1]
##  rvest                    0.3.6      2020-07-25 [1]
##  S4Vectors              * 0.28.0     2020-10-27 [1]
##  scales                 * 1.1.1      2020-05-11 [1]
##  scater                 * 1.18.3     2020-11-08 [1]
##  scuttle                  1.0.0      2020-10-27 [1]
##  SemiPar                * 1.0-4.2    2018-04-16 [1]
##  sessioninfo              1.1.1      2018-11-05 [1]
##  shape                    1.4.5      2020-09-13 [1]
##  shiny                    1.5.0      2020-06-23 [1]
##  SingleCellExperiment   * 1.12.0     2020-10-27 [1]
##  spam                   * 2.5-1      2019-12-12 [1]
##  sparseLDA              * 0.1-9      2016-09-22 [1]
##  sparseMatrixStats        1.2.0      2020-10-27 [1]
##  stringi                  1.5.3      2020-09-09 [1]
##  stringr                * 1.4.0      2019-02-10 [1]
##  SummarizedExperiment   * 1.20.0     2020-10-27 [1]
##  SuppDists                1.1-9.5    2020-01-18 [1]
##  survcomp               * 1.40.0     2020-10-27 [1]
##  survival               * 3.2-7      2020-09-28 [2]
##  survivalROC              1.0.3      2013-01-13 [1]
##  testthat                 3.0.0      2020-10-31 [1]
##  tibble                 * 3.0.4      2020-10-12 [1]
##  tidyr                  * 1.1.2      2020-08-27 [1]
##  tidyselect               1.1.0      2020-05-11 [1]
##  tidyverse              * 1.3.0      2019-11-21 [1]
##  truncnorm                1.0-8      2018-02-27 [1]
##  tweenr                   1.0.1      2018-12-14 [1]
##  usethis                  1.6.3      2020-09-17 [1]
##  vctrs                    0.3.4      2020-08-29 [1]
##  vipor                    0.4.5      2017-03-22 [1]
##  viridis                  0.5.1      2018-03-29 [1]
##  viridisLite              0.3.0      2018-02-01 [1]
##  withr                    2.3.0      2020-09-22 [1]
##  xfun                     0.19       2020-10-30 [1]
##  XML                      3.99-0.5   2020-07-23 [1]
##  xml2                     1.3.2      2020-04-23 [1]
##  xtable                   1.8-4      2019-04-21 [1]
##  XVector                  0.30.0     2020-10-28 [1]
##  yaml                     2.2.1      2020-02-01 [1]
##  zlibbioc                 1.36.0     2020-10-28 [1]
##  source                       
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  Bioconductor                 
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  Bioconductor                 
##  Bioconductor                 
##  Bioconductor                 
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.3)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.3)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.1)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.1)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  Bioconductor                 
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  Github (vqv/ggbiplot@7325e88)
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  Bioconductor                 
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.3)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.3)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.3)               
##  CRAN (R 4.0.3)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.3)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.3)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.3)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.1)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  CRAN (R 4.0.2)               
##  Bioconductor                 
##  CRAN (R 4.0.2)               
##  Bioconductor                 
## 
## [1] /Users/runner/work/_temp/Library
## [2] /Library/Frameworks/R.framework/Versions/4.0/Resources/library</code></pre>
</details>
</div>

<div id="rmd-source-code">---
title: "4. Sparse Singular Value Decomposition"
author: "Lieven Clement"
date: "statOmics, Ghent University (https://statomics.github.io)"
output:
    html_document:
      code_download: true
      theme: cosmo
      toc: true
      toc_float: true
      highlight: tango
      number_sections: true
---

```{r echo = FALSE, warning = FALSE}
library(tidyverse)
library(gridExtra)
```

# Introduction

In high dimensional data the PCs may have unclear interpretation because they depend on all $p$ original variables.

## Example: Toxicogenomics in early drug development

### Background

- Effect of compound on gene expression.

- Insight in action and toxicity of drug in early phase
- Determine activity with bio-assay: e.g. binding afinity of compound to cellwall recepter (target, IC50).
- Early phase:  20 to 50 compounds
- Based on in vitro results one aims to get insight in how to build better compound (higher on-target activity less toxicity.
- Small variantions in moelculair structure lead to variations in BA and gene expression.
- Aim: Build model to predict bio-activity based on gene expression in liver cell line.

### Data

- 30 chemical compounds have been screened for toxicity

- Bioassay data on toxicity screening

- Gene expressions in a liver cell line are profiled for each compound (4000 genes)


```{r}
toxData <- read_csv("https://raw.githubusercontent.com/statOmics/HDA2020/data/toxDataCentered.csv")
svdX <- svd(toxData[,-1])
```

Data is already centered:

```{r}
toxData %>%
  colMeans %>%
  range
```

```{r}
 toxData %>%
  names %>%
  head
```

- First column contains data on Bioassay.
- The higher the score on Bioassay the more toxic the compound
- Other columns contain data on gene expression X1, ... , X4000

### Data exploration

```{r}
toxData %>%
  ggplot(aes(x="",y=BA)) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(position="jitter")
```

```{r}
svdX <- toxData[,-1] %>%
  svd

k <- 2
Vk <- svdX$v[,1:k]
Uk <- svdX$u[,1:k]
Dk <- diag(svdX$d[1:k])
Zk <- Uk%*%Dk
colnames(Zk) <- paste0("Z",1:k)
colnames(Vk) <- paste0("V",1:k)

pca <- Zk %>%
  as.data.frame %>%
  mutate(BA = toxData %>% pull(BA)) %>%
  ggplot(aes(x= Z1, y = Z2, color = BA)) +
  geom_point(size = 3) +
  scale_colour_gradient2(low = "blue",mid="white",high="red") +
  geom_point(size = 3, pch = 21, color = "black")
pca
```

- Scores on the first two principal components (or MDS plot). - Each point corresponds to a compound.
- Color code refers to the toxicity score (higher score more toxic).
- Clear separation between compounds according to toxicity.

---

- Next logic step in a PCA is to interpret the principal components.
- We thus have to assess the loadings.
- We can add a vector for each gene to get a biplot, but this would require plotting 4000 vectors, which would render the plot unreadable.

Alternative graph to look at the many loadings of the first two PCs.

```{r}
grid.arrange(
  Vk %>%
    as.data.frame %>%
    mutate(geneID = 1:nrow(Vk)) %>%
    ggplot(aes(x = geneID, y = V1)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vk[,1]), col = "red") ,
  Vk %>%
    as.data.frame %>%
    mutate(geneID = 1:nrow(Vk)) %>%
    ggplot(aes(x = geneID, y = V2)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vk[,2]), col = "red"),
  ncol=2)
```

- It is almost impossible to interpret the PCs because there are 4000 genes contributing to each PC.

- In an attempt to find the most important genes (in the sense that they drive the interpretation of the PCs), the plots show horizontal reference lines: the average of the loadings, and the average ± twice the standard deviation of the loadings. In between the lines we expects about 95% of the loadings (if they were normally distributed).

- The points outside the band come from the genes that have rather large loadings (in absolute value) and hence are important for the interpretation of the PCs.

- Note, that particularly for the first PC, only a few genes show a markedly large loadings that are negative. This means that an upregulation of these genes will lead to low scores on PC1.
- These genes will very likely play an important role in the toxicity mechanism.
- Indeed, low scores on PC1 are in the direction of more toxicity.

---

## Sparse matrix decomposition

Basic idea of sparse matrix decomposition of an $n \times p$ matrix $\mathbf{X}$}

Find a rank $k$ approximation of $\mathbf{X}$ of the form
\[
\mathbf{X}_k = \sum_{j=1}^k \delta_j \mathbf{u}_j\mathbf{v}_j^t
\]
such that many of the elements in the $\mathbf{v}_j$ and or $\mathbf{u}_j$ are exactly zero.

We focus on the **Penalised Matrix Decomposition** (PMD) method of Witten *et al.* (2009, Biostatistics, 10, 3, 515-534).

When many of the loadings in $\mathbf{v}_j$ are zero, the interpretation of the PC only depends on the features that correspond to the remaining non-zero loadings.

---

# Penalised Matrix Decomposition

## Rank 1 approximation

The rank-1 PMD approximation of the $n \times p$ matrix $\mathbf{X}$ is of the form
\[
\mathbf{X}_1 = \delta_1\mathbf{u}_1\mathbf{v}_1^t,
\]

and it is the solution to the following minimisation problem:
\[
\min_{\delta,u,v} \Vert \mathbf{X} - \delta\mathbf{u}\mathbf{v}^t\Vert_F^2 \text{ subject to }
\delta \geq 0, \Vert \mathbf{u}\Vert_2^2=\Vert \mathbf{v}\Vert_2^2=1, \Vert\mathbf{u}\Vert_1\leq c_1, \Vert \mathbf{v}\Vert_1 \leq c_2
\]

for some user-defined constants $c_1,c_2>0$.

The constraints $\Vert\mathbf{u}\Vert_1\leq c_1$ and $\Vert \mathbf{v}\Vert_1 \leq c_2$ are known as $L_1$ or lasso constraints.


Note,
\[
\Vert\mathbf{u}\Vert_1\ = \sum_{i=1}^n \vert u_i \vert \text{ and }\Vert\mathbf{v}\Vert_1\ = \sum_{i=1}^p \vert v_i\vert.
\]

If $c_1=+\infty$, then the constraint on $\mathbf{u}$ is removed (similar for $c_2$).

---

Note that
\[
  \min_{\delta,u,v} \Vert \mathbf{X} - \delta\mathbf{u}\mathbf{v}^t\Vert_F^2 = \min_{A:\text{rank}(A)=1}  \Vert \mathbf{X} - A\Vert_F^2
\]
with $A=X_1$ (truncated SVD of $\mathbf{X}$, truncated after 1 term). However, the PMD requires a constrained minimisation. The unconstrained solution arises with $c_1=c_2=+\infty$.

Note that $\Vert \mathbf{u}\Vert_2^2=\Vert \mathbf{v}\Vert_2^2=1$ may also be written as $\mathbf{u}^t\mathbf{u}=\mathbf{v}^t\mathbf{v}=1$ (i.e. the vectors are normalised, just like for the singular vectors of the SVD).

- We also applied the $L_1$ constraint to the least-squares estimator of the regression parameters in a linear regression model.

- Effect was that many of the parameter estimates are set exactly to zero.

- Here, the same effect will take place. Many of the elements in $\mathbf{u}$ and $\mathbf{v}$ will be set exactly to zero.

- In particular, the smaller $c_2$, the more elements of $\mathbf{v}$ will be set to zero (shrinkage).

- Similarly, the smaller $c_1$, the more elements of $\mathbf{u}$ will be set to zero.

- The choice of $c_1$ and $c_2$ depends on the objective of the data analysis and will be discussed later.

The solution is known as a rank-1 PMD approximation of $\mathbf{X}$ because $\delta_1\mathbf{u}_1\mathbf{v}_1^t$ is a matrix of rank 1.

---

## Rank-2 approximation

The rank-2 PMD approximation is found as follows:

1. find the rank-1 PMD as before: $\mathbf{X}_1=\delta_1\mathbf{u}_1\mathbf{v}_1^t$;

2. compute
\[
\tilde{\mathbf{X}}_2 = \mathbf{X}-\mathbf{X}_1,
\]

and compute the rank-1 PMD of $\tilde{\mathbf{X}}_2$, which gives $\delta_2, \mathbf{u}_2, \mathbf{v}_2$;

3. the rank-2 PMD of $\mathbf{X}$ is then given by
\[
\mathbf{X}_2 = \delta_1\mathbf{u}_1\mathbf{v}_1^t+\delta_2\mathbf{u}_2\mathbf{v}_2^t.
\]

Note that the $\mathbf{u}$'s and $\mathbf{v}$'s are not necessarily orthogonal (only if $c_1, c_2=+\infty$).

---

## Rank-k approximation

The rank-$k$ PMD approximation is found as follows:

1. find the rank-1 PMD as before: $\tilde{\mathbf{X}}_1=\delta_1\mathbf{u}_1\mathbf{v}_1^t$;

2. set $j=1$;

3. compute \[
\tilde{\mathbf{X}}_{j+1} = \mathbf{X}-\tilde{\mathbf{X}}_1-\cdots - \tilde{\mathbf{X}}_j,
\] and compute the rank-1 PMD of $\tilde{\mathbf{X}}_{j+1}$, which gives $\delta_{j+1}, \mathbf{u}_{j+1}, \mathbf{v}_{j+1}$;

4. if $j<k$, increase $j$ with one and go back to step 3; otherwise go to step 5;

5. the rank-$k$ PMD of $\mathbf{X}$ is then given by
   \[
     \mathbf{X}_k = \sum_{j=1}^k \delta_j\mathbf{u}_j\mathbf{v}_j^t.
   \]

Without the $L_1$ constraints, this algorithm gives the ordinary SVD.

As before the $\delta_j$s, $\mathbf{u}_j$s and $\mathbf{v}_j$s are stacked into matrices $\mathbf{D}_k$, $\mathbf{U}_k$ and $\mathbf{V}_k$: $\mathbf{X}_k = \mathbf{U}_k\mathbf{D}_k\mathbf{V}_k^t$.


Note that the $\mathbf{u}$ vectors are generally not orthogonal (i.e. $\mathbf{u}_i^t\mathbf{u}_j\neq 0$ for $i\neq j$) (the same holds for the $\mathbf{v}$ vectors).

---

## Sparse PCA via PMD

Just like the SVD can be used to find the PCA solution, the PMD can be used to find a sparse PCA solution (SPC).

For a PCA, it is only important to have sparse loadings (in the $\mathbf{v}$s); the scores (involving the $\mathbf{u}$s) need not be sparse.

The rank-1 solutions thus satisfy
\[
  \min_{\delta,u,v} \Vert \mathbf{X} - \delta\mathbf{u}\mathbf{v}^t\Vert_F^2 \text{ subject to }
  \delta \geq 0, \Vert \mathbf{u}\Vert_2^2=\Vert \mathbf{v}\Vert_2^2=1, \vert \mathbf{v}\vert \leq c
\]
(i.e. no sparsity constraint on $\mathbf{u}$).

Hence the PCA scores
\[
 \mathbf{Z}_k = \mathbf{U}_k\mathbf{D}_k
\]
relate to $\mathbf{X}_k$ through the sparse loadings in $\mathbf{V}_k$.


---

It can be shown that the SPC solution (rank-1) also maximises
\[
  \mathbf{v}^t\mathbf{X}^t\mathbf{X}\mathbf{v} \text{ subject to }
  \Vert \mathbf{u}\Vert_2^2=\Vert \mathbf{v}\Vert_2^2=1, \Vert \mathbf{v}\Vert_1 \leq c
\]
(i.e. it maximises the variance of the PCs, subject to sparsity constraint on loadings).


From these two slides we may conclude that we can use the PMD as the basis of a sparse PCA, and that the interpretation is as before in terms of maximising variance. Only now we hope that only a few features will give a non-zero loading in the $\mathbf{v}$-vectors. Biplots may again be constructed (with fewer vectors to be plotted).

---

# Toxicogenomics analysis with a sparse PCA

## SPCA

```{r}
library(PMA)

spcX <- SPC(
  toxData[,-1] %>%
    as.matrix,
  K=30,
  sumabsv = 5)

par(mfrow=c(1,2))
barplot(spcX$d^2,ylab="eigenvalue",cex.lab=1.5)
barplot(spcX$prop.var.explained,
     ylab="cummulative variance",cex.lab=1.5)

```


- The plot on the left shows the eigenvalues (squared singular values) from the PMD.

  - This suggests that the first one or two are the most important, but all eigenvalues remain substantially larger than zero.

- The figure on the right shows the cumulative percentage of variance retained in the PCs.

    - These percentages are now no longer calculated as before as ratio's of sums of eigenvalues.
    - A more complicated calculation is required now, because the PCs are no longer uncorrelated with one another.
    - Thus part of the information (variance) of one PC is shared with part of the information of another PCs (i.e. non-zero correlation or covariance).

## Loading

Loadings of the first two PCs, with constraint $\Vert\mathbf{v}\Vert_1 < 5$.

```{r}
k <- 2
Vk <- spcX$v[,1:k]
colnames(Vk) <- paste0("V",1:k)

grid.arrange(
  Vk %>%
    as.data.frame %>%
    mutate(geneID = 1:nrow(Vk)) %>%
    ggplot(aes(x = geneID, y = V1)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vk[,1]), col = "red") ,
  Vk %>%
    as.data.frame %>%
    mutate(geneID = 1:nrow(Vk)) %>%
    ggplot(aes(x = geneID, y = V2)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vk[,2]), col = "red"),
  ncol=2)
```

These plots show the loadings of the first two sparse PCs. Many of the loadings are now exactly equal to zero. Only few genes show large loadings.

## Change constraint

We now repeat the analysis, but with the constraint $\Vert \mathbf{v} \Vert_1 < 1$. This gives a much sparser solution with only two genes with non-zero loadings in each dimension.

```{r}

spcX1 <- SPC(
  toxData[,-1] %>%
    as.matrix,
  K=30,
  sumabsv = 1)

k <- 2
Vk <- spcX1$v[,1:k]
colnames(Vk) <- paste0("V", 1:k)

grid.arrange(
  Vk %>%
    as.data.frame %>%
    mutate(geneID = 1:nrow(Vk)) %>%
    ggplot(aes(x = geneID, y = V1)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vk[,1]), col = "red") ,
  Vk %>%
    as.data.frame %>%
    mutate(geneID = 1:nrow(Vk)) %>%
    ggplot(aes(x = geneID, y = V2)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vk[,2]), col = "red"),
  ncol=2)
```

---

We now repeat the analysis, but with the constraint $\Vert \mathbf{v} \Vert_1 < 10$. This gives a much less sparse solution with several genes with non-zero loadings.

```{r}
spcX10 <- SPC(
  toxData[,-1] %>%
    as.matrix,
  K=30,
  sumabsv = 10)

k <- 2
Vk <- spcX10$v[,1:k]
colnames(Vk) <- paste0("V",1:k)

grid.arrange(
  Vk %>%
    as.data.frame %>%
    mutate(geneID = 1:nrow(Vk)) %>%
    ggplot(aes(x = geneID, y = V1)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vk[,1]), col = "red") ,
  Vk %>%
    as.data.frame %>%
    mutate(geneID = 1:nrow(Vk)) %>%
    ggplot(aes(x = geneID, y = V2)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vk[,2]), col = "red"),
  ncol=2)
```

---

## Variance explained

Variance explained for $c=1,5,10$ and $+\infty$.

```{r}
data.frame(
  PCs = 1:30,
  "c1" = spcX1$prop.var.explained,
  "c5" = spcX$prop.var.explained,
  "c10" = spcX10$prop.var.explained,
  "conventional" = cumsum(svdX$d^2)/sum(svdX$d^2)
  ) %>%
  gather("method","varExpl",-1) %>%
  ggplot(aes(x = PCs, y = varExpl, color = method)) +
  geom_line() +
  geom_point() +
  ylab("Variance explained (%)")
```

This graph illustrates that the percentages of variance explained by the PCs are smaller for sparser solutions. Thus in choosing an appropriate penalty parameter $c_2$ one should not only look at the sparseness of the solution, but  also at the percentage of variance contained in the first few PCs. If this percentage is very small, then the graphs of the scores (or biplots) are no longer very informative (i.e. they will not tell the full story). Thus a compromise must be looked for.

```{r}
library(latex2exp)

Uk5 <- spcX$u[,1:k]
Dk5 <- diag(spcX$d[1:k])
Zk5 <- Uk5%*%Dk5
colnames(Zk5) <- paste0("Z",1:k)

spc5 <- Zk5 %>%
  as.data.frame %>%
  mutate(BA = toxData %>% pull(BA)) %>%
  ggplot(aes(x= Z1, y = Z2, color = BA)) +
  geom_point(size = 3) +
  scale_colour_gradient2(low = "blue",mid="white",high="red") +
  geom_point(size = 3, pch = 21, color = "black") +
  ggtitle(TeX("$c_2 = 5$"))

Uk1 <- spcX1$u[,1:k]
Dk1 <- diag(spcX1$d[1:k])
Zk1 <- Uk1%*%Dk1
colnames(Zk1) <- paste0("Z",1:k)

spc1 <- Zk1 %>%
  as.data.frame %>%
  mutate(BA = toxData %>% pull(BA)) %>%
  ggplot(aes(x= Z1, y = Z2, color = BA)) +
  geom_point(size = 3) +
  scale_colour_gradient2(low = "blue",mid="white",high="red") +
  geom_point(size = 3, pch = 21, color = "black") +
  ggtitle(TeX("$c_2 = 1$"))

Uk10 <- spcX10$u[,1:k]
Dk10 <- diag(spcX10$d[1:k])
Zk10 <- Uk10%*%Dk10
colnames(Zk10) <- paste0("Z",1:k)

spc10 <- Zk10 %>%
  as.data.frame %>%
  mutate(BA = toxData %>% pull(BA)) %>%
  ggplot(aes(x= Z1, y = Z2, color = BA)) +
  geom_point(size = 3) +
  scale_colour_gradient2(low = "blue",mid="white",high="red") +
  geom_point(size = 3, pch = 21, color = "black") +
  ggtitle(TeX("$c_2 = 10$"))

grid.arrange(spc1, spc5, spc10, pca + ggtitle(TeX("$c_2 = \\infty")), ncol = 2)
```

Finally we show the plots of the scores on the first two sparse PCs for $c_2=1$ (top-left), $c_2=5$ (top-right) and $c_2=10$ (bottom-left) and $c_2=\infty$ (bottom-right, convention SVD).

Note, that the most sparse solution (top-left) shows a strong positive correlation between the first two PCs (as a consequence of the loss of orthogonality of the singular vectors).

All graphs succeed quite well in separating the toxic from the non-toxic compounds based on their gene expression.

- Hence, we may expect that the two genes identified from the most sparse solution play an important role in the development of toxic effects.

- Note, however, that our data analysis did not provide a scientific proof for this conclusion: (sparse) PCA is only an exploratory or descriptive data analysis method that helps the data-analyst to gain insight into the data.

- The fact that we found a few genes that appear to be associated with toxicity is of course a nice result.

- It is important to understand that the toxicity outcome was not used in the (sparse) PCA. We only used the toxicity to color the points in the graphs. PCA is an unsupervised method.

# Acknowledgement {-}

- Olivier Thas for sharing his materials of Analysis of High Dimensional Data 2019-2020, which I used as the starting point for this chapter.  

# Session info {-}

<details><summary>Session info</summary>

```{r session_info, echo=FALSE, cache=FALSE}
Sys.time()
sessioninfo::session_info()
```

</details>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("sparseSvd.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
